src/main/java/org/elasticsearch/index/query/QueryBuilder.java

```diff
@@ -45,4 +45,10 @@
      * @throws IOException
      */
     Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException;
+
+    /**
+     * Validate the query.
+     * @throws ElasticsearchException if the query is not valid, e.g. fields that are needed to create the lucene query are missing.
+     */
+    void validate() throws ElasticsearchException;
```

Comment:
I think we should stick to the current validate api for now, although it does feel weird that we don't actually throw exception. What happens is that the listener gets notified of the exception in `TransportAction#execute`. We can't yet hook into the validate mechanism as the `SearchRequest` only contains a `source` bytes array in json format, but at the end of the refactoring we should have different elements that hold the different parts of a search request, among which the query which can be validated as part of `SearchRequest#validate`. This is to me the only way to make sure that validation happens whenever needed.